### PR TITLE
eip6110: Queue deposit requests and apply them during epoch processing

### PR DESCRIPTION
**Pull request by mkalinin**
### Description
Introduces majority of things described in https://github.com/ethereum/consensus-specs/pull/3689#issuecomment-2112987260, specifically:
* Queues full deposit request objects and process them on the epoch processing, rate limits a number of deposit signature verifications per epoch by the max number of deposit requests allowed to be processed (`MAX_PENDING_DEPOSITS_PER_EPOCH`)
* Finalizes deposit request position in the queue before applying it to the state. The outcome is that no new validator is created before the corresponding deposit is finalized hence no fork aware `(pubkey, index)` cache is required.
* Eth1 bridge deposits are processed as before this change, the corresponding entry with the entire balance is created in the `pending_deposits` queue to ensure the deposit will pass through the activation churn.
* No deposit request is processed until Eth1 bridge covers the gap during the transition period. This prevents the scenario when deposit request can easily front-run the Eth1 bridge deposit because of the large follow distance and create validator with the same pubkey but malicious withdrawal credentials.
* Refactors `process_pending_deposits` by moving deposit applying to `apply_pending_deposit`

Other than the above, the logic of the previous `process_pending_balance_deposits` function remains.

~~Additionally, moves `switch_to_compounding_validator` to `process_deposit_request`~~. Removes `switch_to_compounding_validator` from the deposit flow in favour of https://github.com/ethereum/consensus-specs/pull/3918.

### Points of discussion and open questions
* Due to refactor that facilitates spec reading there are two places where `validatorIndex` is looked up by the `pubkey` (`process_pending_deposits ` and `apply_pending_deposit` functions), this happens for each pending deposit that is being processed. There is a less readable workaround where `validatorIndex` is resolved once, but I am not sure if it is necessary because client implementations are likely to use cache in both cases.
* `MAX_PENDING_DEPOSITS_PER_EPOCH_PROCESSING` is set to `16`, should it be more? The activation churn is limited by `256 ETH` which would mean 256 signature verifications in the worst case. The most optimal scenario of an attack resulting in 256 sig verifications would entail creation of `128` 1-ETH validators (the first deposit would be to create a validator with eth1 creds, the second would be to switch the validator to compounding creds); it seems that employing such an attack is quite discouraging because it requires decent amount of ETH with no substantial gain. Should we even have this limitation if the churn could do the limit?
* This PR does not explore an optimisation with two queues that was discussed previously, it is planned to be a part of a separate PR once this PR gets merged.

### Testing

_Many thanks to @james-prysm for helping to write and adjust the tests!_ 

* [x] `apply_pending_deposits` checks
* [x] Deposit transition period checks -- the logic has changed and part of it is in `process_pending_deposit` 
* [x] Repeat `test_process_deposit_requests` scenarious in application to `process_pending_deposit`
* [x] Finalized deposit position check
* [x] `MAX_PENDING_DEPOSITS_PER_EPOCH_PROCESSING` check
* [x] Revisit `test_process_pending_deposit` (`is_deposit_applied`, `is_churn_limit_reached` etc)

Link to a more comprehensive test plan: https://hackmd.io/mqV7TaFyTq--vEMcB5msJw

Supersedes https://github.com/ethereum/consensus-specs/pull/3689

### CODE REVIEW

File: specs/electra/beacon-chain.md
Comment (dapplion) [inline specs/electra/beacon-chain.md:641]: From @mkalinin this line is motivated by
- https://github.com/ethereum/consensus-specs/pull/3776

Comment (mkalinin) [reply specs/electra/beacon-chain.md:641]: The idea is to postpone deposit processing for an exiting validator until it has fully withdrawn. And then do the top-up without affecting WS period.

**Comment by james-prysm** [comment]
> Due to refactor that facilitates spec reading there are two places where validatorIndex is looked up by the pubkey (get_activation_churn_consumption and apply_pending_deposit functions), this happens for each pending deposit that is being processed. There is a less readable workaround where validatorIndex is resolved once, but I am not sure if it is necessary because client implementations are likely to use cache in both cases.

does this mean that the cache should continue to be used as opposed to the previous pr that looked for its removal in https://github.com/ethereum/consensus-specs/pull/3689

File: specs/electra/beacon-chain.md
Comment (james-prysm) [inline specs/electra/beacon-chain.md:932]: later in the code when the churn_limit_reached is checked could it be possible for the balance to be negative if `churn_consumption == 0` from `get_activation_churn_consumption` 
that would mean that process_amount can be more than available_for_processing. 

hope i read that correctly

Comment (mkalinin) [reply specs/electra/beacon-chain.md:932]: It can’t be possible because of this check. If `processed_amount` is about to get higher than `available_for_processing` the loop gets aborted before `processed_amount` is increased down the line.

File: specs/electra/beacon-chain.md
Comment (james-prysm) [inline specs/electra/beacon-chain.md:957]: I think because we start at index 0 for next_deposit_index, `next_deposit_index > MAX_PENDING_DEPOSITS_PER_EPOCH_PROCESSING` only triggers when  next_deposit_index has incremented MAX_PENDING_DEPOSITS_PER_EPOCH_PROCESSING+2 times, which i'm not sure was the intended. maybe this needs to be
`next_deposit_index > MAX_PENDING_DEPOSITS_PER_EPOCH_PROCESSING-1` to account for the 0 index?

Comment (mkalinin) [reply specs/electra/beacon-chain.md:957]: Good catch! It should rather be:
```suggestion
        if next_deposit_index >= MAX_PENDING_DEPOSITS_PER_EPOCH_PROCESSING:
```

File: specs/electra/beacon-chain.md
Comment (james-prysm) [inline specs/electra/beacon-chain.md:893]: oh this is greatly simplified

File: specs/electra/beacon-chain.md
Comment (mkalinin) [reply specs/electra/beacon-chain.md:946]: Yes, thank you!

File: specs/electra/beacon-chain.md
Comment (james-prysm) [inline specs/electra/beacon-chain.md:1285]: the comment that it was modified should be removed right?

File: specs/electra/beacon-chain.md
Comment (james-prysm) [inline specs/electra/beacon-chain.md:970]: can we combine 
```
state.pending_deposits = state.pending_deposits[next_deposit_index:]
```
and

```
state.pending_deposits += deposits_to_postpone
```
?? or is this for readability

File: specs/electra/beacon-chain.md
Comment (james-prysm) [inline specs/electra/beacon-chain.md:1285]: what about
```
if pubkey not in validator_pubkeys:
  if I!s_valid_deposit_signature(pubkey, withdrawal_credentials, amount, signature):
    break
 add_validator_to_registry(state, pubkey, withdrawal_credentials, Gwei(0))
state.pending_deposits.append(PendingDeposit(
                pubkey=pubkey,
                withdrawal_credentials=withdrawal_credentials,
                amount=amount,
                signature=signature,
                slot=GENESIS_SLOT,
            ))
```

**Review by ralexstokes — APPROVED** [review]
great work! 

I think this is the next best step here

I'd also want to see performance numbers if we just rely on the churn limit instead of having an additional constant

and I'd also lean against the next suggested step around managing the queue in a more sophisticated way unless its obvious we need it for performance reasons

**Comment by mkalinin** [comment]
> I'd also want to see performance numbers if we just rely on the churn limit instead of having an additional constant

The most optimal DoS strategy seems to be the following:
1) Create validator with 1 ETH balance and ETH1 creds
2) Top-up the validator with 1 ETH and COMPOUNDING creds

Each of these steps yields signature verification. Current churn limit allows to create 128 validators using this strategy, that is 256 signature verifications at the epoch processing boundary (256ms). The cost of the attack is locked ETH and some amount of ETH required to activate those validators by topping them up with 30 ETH and withdraw them. In the worst case it will take yet another 15 epochs to withdraw all malicious validators. Potential outcome and the cost makes this attack ineffective to my perspective. If the churn limit is increased then the outcome may become worse but the cost will also increase.

File: presets/mainnet/electra.yaml
Comment (ensi321) [inline presets/mainnet/electra.yaml:50]: I understand this limit is specifically for epoch processing. Does it need to emphasize the epoch processing part? Asking because there's no precedent on presets having `_PER_EPOCH_PROCESSING` suffix but `_PER_EPOCH`

File: specs/electra/beacon-chain.md
Comment (mkalinin) [reply specs/electra/beacon-chain.md:713]: The protocol does not accept a block proposed in the `GENESIS_SLOT` because of the following check in the `process_block_header`:
```python
assert block.slot > state.latest_block_header.slot
```

File: specs/electra/beacon-chain.md
Comment (mkalinin) [reply specs/electra/beacon-chain.md:1373]: Yes, these functions are going to be deprecated and I think that we will be able/should update the genesis state generation if possible and get rid of these functions

File: specs/electra/beacon-chain.md
Comment (mkalinin) [reply specs/electra/beacon-chain.md:1577]: `deposit.index` is the index of the deposit record in the smart contract, it can’t be used to derive the `validator_index` anyhow

File: specs/electra/beacon-chain.md
Comment (james-prysm) [reply specs/electra/beacon-chain.md:1296]: oops missed this comment, yeah just readability, as the appending logic is duplicated as far as I could tell

**Review by james-prysm — APPROVED** [review]

**Review by james-prysm — APPROVED** [review]
just signalling support , but also waiting for ` switch_to_compounding` piggy backing off consolidations

**Comment by ralexstokes** [comment]
AIUI we would merge in https://github.com/ethereum/consensus-specs/pull/3918 which will simplify this PR a bit further

does this align with your understanding @mkalinin ?

**Comment by mkalinin** [comment]
> AIUI we would merge in #3918 which will simplify this PR a bit further
> 
> does this align with your understanding @mkalinin ?

yes, the plan is to wait for that PR and then merge this one

**Comment by Scutua** [comment]
Hello

**Comment by mkalinin** [comment]
`switch_to_compounding_validator` is removed from the deposit flow in favour of https://github.com/ethereum/consensus-specs/pull/3918

File: specs/electra/beacon-chain.md
Comment (terencechain) [inline specs/electra/beacon-chain.md:1321]: `apply_deposit` doesn't seem to be used anymore in Electra. Am I missing anything?

Comment (lucassaldanha) [reply specs/electra/beacon-chain.md:1321]: It is still used in `process_deposit`. I don't think we are getting rid of yet?

Comment (terencechain) [reply specs/electra/beacon-chain.md:1321]: Gotcha. Then we should add `process_deposit` here in this spec as *modified*

Comment (mkalinin) [reply specs/electra/beacon-chain.md:1341]: Done!

File: tests/core/pyspec/eth2spec/test/electra/block_processing/test_process_deposit_request.py
Comment (jtraglia) [inline tests/core/pyspec/eth2spec/test/electra/block_processing/test_process_deposit_request.py:136]: We should add an assert that `deposit_request.index` is different than `initial_start_index`:
```suggestion
    assert initial_start_index != deposit_request.index
    state.deposit_requests_start_index = initial_start_index
```

If they're the same, the test doesn't tell us anything.

File: tests/core/pyspec/eth2spec/test/electra/block_processing/test_process_deposit_request.py
Comment (jtraglia) [inline tests/core/pyspec/eth2spec/test/electra/block_processing/test_process_deposit_request.py:178]: This insightful comment is no longer in `test_process_deposit_request_top_up_invalid_sig`.

Is this still accurate?

**Review by jtraglia — COMMENTED** [review]
This PR is looking good. I've reviewed the spec changes and found no issues, but reviewing the tests is fairly difficult. From what I've seen they look quite good too. I'll try to review them more later, but I'm inclined to trust they are doing the right thing.

File: tests/core/pyspec/eth2spec/test/electra/block_processing/test_process_deposit_request.py
Comment (mkalinin) [reply tests/core/pyspec/eth2spec/test/electra/block_processing/test_process_deposit_request.py:178]: It is a bit tricky. This is still true, but the test that checks that is `test_apply_pending_deposit_incorrect_sig_top_up` as there is no sig validations on deposit request processing. Deposit request tests ensures that pending deposits are created with invalid deposit request data of different kind, i.e. all validations are _deferred_ to `apply_pending_deposit`.

File: specs/electra/beacon-chain.md
Comment (ppopth) [reply specs/electra/beacon-chain.md:1328]: I think the status quo is more readable.

File: specs/electra/beacon-chain.md
Comment (ppopth) [inline specs/electra/beacon-chain.md:1371]: It seems that `process_deposit` is not modified at all. I think we don't have to put it in the Electra spec. I don't think changes on `apply_deposit` make it qualified for this.

File: specs/electra/beacon-chain.md
Comment (ppopth) [inline specs/electra/beacon-chain.md:661]: What is the consequence if I set this to `state.slot` instead?

Comment (ppopth) [inline specs/electra/beacon-chain.md:661]: I think we should define meaningful constants for both `bls.G2_POINT_AT_INFINITY` and `GENESIS_SLOT`. How about aliasing them as
- `UNSET_SIGNATURE` and
- `UNSET_DEPOSIT_SLOT`

By seeing only the point at infinity and the genesis slot, the readers wouldn't know what they mean.

Comment (mkalinin) [reply specs/electra/beacon-chain.md:661]: Then the excess balance will wait for when `state.slot` is finalized before being applied even in the case when there is enough activation churn to let the balance in. The `GENESIS_SLOT` is used to bypass that finality check as that slot is always finalized

**Comment by jtraglia** [comment]
> > AIUI we would merge in #3918 which will simplify this PR a bit further
> > does this align with your understanding @mkalinin ?
> 
> yes, the plan is to wait for that PR and then merge this one

@mkalinin We've merged #3918. Any additional changes you'd like to make?

**Comment by jtraglia** [comment]
After merging in dev, the `test_pending_consolidation_future_epoch` test fails. I probably messed something up but it's not clear to me what is wrong. Do you know?

File: tests/core/pyspec/eth2spec/test/electra/epoch_processing/test_apply_pending_deposit.py
Comment (hwwhww) [inline tests/core/pyspec/eth2spec/test/electra/epoch_processing/test_apply_pending_deposit.py:1]: > `electra/epoch_processing/test_apply_pending_deposit.py`

It's not a typical "epoch processing" test format test suite. We need to either (1) define a new test format for it or (2) rename this helper to `process_pending_deposit` (or another name to avoid confusion with `process_pending_deposits`). The latter is more straightforward.

File: tests/core/pyspec/eth2spec/test/helpers/deposits.py
Comment (hwwhww) [inline tests/core/pyspec/eth2spec/test/helpers/deposits.py:368]: is it needed?

File: tests/core/pyspec/eth2spec/test/helpers/deposits.py
Comment (hwwhww) [inline tests/core/pyspec/eth2spec/test/helpers/deposits.py:411]: is it needed?

File: tests/generators/epoch_processing/main.py
Comment (hwwhww) [inline tests/generators/epoch_processing/main.py:44]: may add `apply_pending_balance_deposit` after it is renamed

**Review by jtraglia — APPROVED** [review]

File: tests/core/pyspec/eth2spec/test/helpers/deposits.py
Comment (mkalinin) [reply tests/core/pyspec/eth2spec/test/helpers/deposits.py:368]: Might be not as these methods are called from post electra tests only

**Pull request merged**
